src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/InheritDocTaglet.java

```diff
@@ -70,25 +73,72 @@ public InheritDocTaglet() {
      */
     private Content retrieveInheritedDocumentation(TagletWriter writer,
                                                    ExecutableElement method,
-                                                   DocTree inheritDoc,
+                                                   InheritDocTree inheritDoc,
                                                    boolean isFirstSentence) {
         Content replacement = writer.getOutputInstance();
         BaseConfiguration configuration = writer.configuration();
         Messages messages = configuration.getMessages();
         Utils utils = configuration.utils;
         CommentHelper ch = utils.getCommentHelper(method);
-        var path = ch.getDocTreePath(inheritDoc).getParentPath();
+        DocTreePath inheritDocPath = ch.getDocTreePath(inheritDoc);
+        var path = inheritDocPath.getParentPath();
         DocTree holderTag = path.getLeaf();
+
+        ExecutableElement src = null;
+        // 1. Does @inheritDoc specify a type?
+        if (inheritDoc.getSupertype() != null) {
+            // 2. Can we find that type?
+            var supertype = (TypeElement) ch.getReferencedElement(inheritDoc.getSupertype());
+            if (supertype == null) {
+                messages.error(inheritDocPath, "doclet.inheritDocBadSupertype");
+                return replacement;
+            }
+            // 3. Does that type have a method that this method overrides?
+            // Skip a direct check that the type that declares this method is a subtype
+            // of the type that @inheritDoc specifies. Do the "overrides" check for
+            // individual methods. Not only will such a check find the overridden
+            // method, but it will also filter out self-referring inheritors
+            // (S is the same type as the type that contains {@inheritDoc S})
+            // due to irreflexivity of the "overrides" relationship.
+            //
+            // This way we do more work in erroneous case, but less in the typical
+            // case. We don't optimize for the former.
+            VisibleMemberTable visibleMemberTable = configuration.getVisibleMemberTable(supertype);
+            List<Element> methods = visibleMemberTable.getAllVisibleMembers(VisibleMemberTable.Kind.METHODS);
+            for (Element e : methods) {
+                ExecutableElement m = (ExecutableElement) e;
+                if (configuration.utils.elementUtils.overrides(method, m, (TypeElement) method.getEnclosingElement())) {
```

Comment:
Suggestion for eventual future cleanup (not this PR): reduce need for chain fields like`config.utils.elementUtils`